src/Microsoft.Identity.Web/TokenAcquisition.cs

```diff
@@ -312,36 +338,56 @@ private IConfidentialClientApplication BuildConfidentialClientApplication()
 
-            if (!applicationOptions.Instance.EndsWith("/"))
-                applicationOptions.Instance += "/";
-
-            string authority ;
-            IConfidentialClientApplication app = null;
-
-            if (microsoftIdentityOptions.IsB2C)
+           if (string.IsNullOrEmpty(applicationOptions.Instance) &&
+                !string.IsNullOrEmpty(microsoftIdentityOptions.Instance))
             {
-                authority = $"{applicationOptions.Instance}tfp/{microsoftIdentityOptions.Domain}/{microsoftIdentityOptions.DefaultUserFlow}";
-                app = ConfidentialClientApplicationBuilder
-                    .CreateWithApplicationOptions(applicationOptions)
-                    .WithRedirectUri(currentUri)
-                    .WithB2CAuthority(authority)
-                    .Build();
+                if (!microsoftIdentityOptions.Instance.EndsWith("/"))
+                    microsoftIdentityOptions.Instance += "/";
+                instance = microsoftIdentityOptions.Instance;
             }
             else
             {
-                authority = $"{applicationOptions.Instance}{applicationOptions.TenantId}/";
-                app = ConfidentialClientApplicationBuilder
-                    .CreateWithApplicationOptions(applicationOptions)
-                    .WithRedirectUri(currentUri)
-                    .WithAuthority(authority)
-                    .Build();
+                if (!applicationOptions.Instance.EndsWith("/"))
+                    applicationOptions.Instance += "/";
+                instance = applicationOptions.Instance;
             }
 
-            // Initialize token cache providers
-            _tokenCacheProvider?.InitializeAsync(app.AppTokenCache);
-            _tokenCacheProvider?.InitializeAsync(app.UserTokenCache);
+            string authority;
+            IConfidentialClientApplication app;
 
-            return app;
+            try
+            {
+                if (microsoftIdentityOptions.IsB2C)
+                {
+                    authority = $"{instance}tfp/{microsoftIdentityOptions.Domain}/{microsoftIdentityOptions.DefaultUserFlow}";
+                    app = ConfidentialClientApplicationBuilder
+                        .CreateWithApplicationOptions(applicationOptions)
+                        .WithRedirectUri(currentUri)
+                        .WithB2CAuthority(authority)
+                        .Build();
+                }
+                else
+                {
+                    authority = $"{instance}{applicationOptions.TenantId}/";
+
+                    app = ConfidentialClientApplicationBuilder
+                        .CreateWithApplicationOptions(applicationOptions)
+                        //.WithRedirectUri(currentUri)
+                        .WithAuthority(authority)
+                        .Build();
+                }
+
+                // Initialize token cache providers
+                _tokenCacheProvider?.InitializeAsync(app.AppTokenCache);
```

Comment:
Aren't you getting warnings about these lines? As a general rule of thumb, you should always (well 99% of time) `await` async operations. Otherwise you have a race condition here - you return an app and can use it to fetch tokens, but the token cache provider might not have finished initialzing. 

Also, because you don't await, you will ignore possible exceptions in the InitializeAsync.

Reply:
This will of course mean that this private method also becomes Async.

Reply:
i wasn't getting warnings but have fixed all of these. good catch.

---
In reply to: [385595862](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385595862) [](ancestors = 385595862)